sw/otbn/code-snippets/p384.s

```diff
@@ -1372,52 +1372,300 @@ scalar_mult_int_p384:
 
   /* set pointer to blinding parameter */
-  la       x9, dptr_rnd
-  lw       x9, 0(x9)
+  la        x9, dptr_rnd
+  lw        x9, 0(x9)
 
   /* set dmem pointer to domain parameter b */
-  la       x28, p384_b
+  la        x28, p384_b
 
   /* set dmem pointer to scratchpad */
-  la       x30, scratchpad
+  la        x30, scratchpad
 
   /* load domain parameter p (modulus)
-     [w13, w12] = p = dmem[dptr_p] */
-  li       x2, 12
-  la       x3, p384_p
-  bn.lid   x2++, 0(x3)
-  bn.lid   x2++, 32(x3)
+     [w13, w12] = p = dmem[p384_p] */
+  li        x2, 12
+  la        x3, p384_p
+  bn.lid    x2++, 0(x3)
+  bn.lid    x2++, 32(x3)
 
   /* load Barrett constant u for modulus p
-     [w15, w14] = u_p = dmem[dptr_u_p] */
-  li       x2, 14
-  la       x3, p384_u_p
-  bn.lid   x2++, 0(x3)
-  bn.lid   x2++, 32(x3)
+     [w15, w14] = u_p = dmem[p384_u_p] */
+  li        x2, 14
+  la        x3, p384_u_p
+  bn.lid    x2++, 0(x3)
+  bn.lid    x2++, 32(x3)
 
   /* load domain parameter n (order of base point)
-     [w11, w10] = p = dmem[dptr_n] */
-  li       x2, 10
-  la       x3, p384_n
-  bn.lid   x2++, 0(x3)
-  bn.lid   x2++, 32(x3)
+     [w11, w10] = n = dmem[p384_n] */
+  li        x2, 10
+  la        x3, p384_n
+  bn.lid    x2++, 0(x3)
+  bn.lid    x2++, 32(x3)
 
   /* init all-zero reg */
-  bn.xor   w31, w31, w31
+  bn.xor    w31, w31, w31
 
-  jal      x1, scalar_mult_int_p384
+  jal       x1, scalar_mult_int_p384
+
+  ret
+
+
+/**
+ * Variable time modular multiplicative inverse computation
+ *
+ * returns x_inv = x^-1 mod m
+ *
+ * This routine computes the modular multiplicative inverse for any x < m in
+ * the finite field GF(m) where m is prime.
+ *
+ * For inverse computation, Fermat's little theorem is used, i.e.
+ * we compute x^-1 = x^(m-2) mod m.
+ * For exponentiation we use a standard, variable time (!) square and multiply
+ * algorithm.
+ *
+ * This routine is mainly intended to be used for inversion of scalars in
+ * context of the P-384 curve. In theory, it can be used with any 384-bit
+ * modulus m with a corresponding 385-bit Barrett constant u,
+ * where u[383:192] = 0.
+ *
+ * Note: When used for P-384 scalar inversion, the routine will need 672 calls
+ * to the multiplication routine. By using an adder chain this could be reduced
+ * to ~433 multiplications, however, for the cost of a significant increase of
+ * codes size.
+ *
+ * Note: This routine runs in variable time w.r.t. the modulus. It should only
+ * be used with a non-secret modulus.
+ *
+ * @param[in]  [w13, w12]: m, 384 bit modulus
+ * @param[in]  [w15, w14]: u[383:0], lower 384 bit of pre-computed Barrett
+ *                          constant corresponding to modulus m
+ * @param[in]  [w30, w29]: x, 384 bit operand
+ * @param[in]  w31, all-zero
+ * @param[out] [w17, w16]: x_inv, modular multiplicative inverse
+ *
+ * Flags: Flags when leaving this subroutine depend on a potentially discarded
+ *        value and therefore are not usable after return.
```

Comment:
Could you rephrase that sentence, I'm not sure I understand it. How can a flag depend on another value? Do you mean the flag has no meaning beyond the scope of this subroutine?

Reply:
It is often the case that flags depend on the result of last computation in the subroutine (and by this on the result), so one could think of using the M/L/Z flags in a follow-up computation. But often this final computation is a conditional subtraction that is executed in any case (for the sake of constant time) but then potentially discarded. I agree, your suggestion (no meaning beyond the scope...) is clearer. Changed this here and at other places as well.